cajita/src/Cajita_ParticleGridDistributor.hpp

```diff
@@ -0,0 +1,390 @@
+/****************************************************************************
+ * Copyright (c) 2018-2021 by the Cabana authors                            *
+ * All rights reserved.                                                     *
+ *                                                                          *
+ * This file is part of the Cabana library. Cabana is distributed under a   *
+ * BSD 3-clause license. For the licensing terms see the LICENSE file in    *
+ * the top-level directory.                                                 *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+
+#ifndef CABANA_PARTICLEGRIDDISTRIBUTOR_HPP
+#define CABANA_PARTICLEGRIDDISTRIBUTOR_HPP
+
+#include <Cabana_DeepCopy.hpp>
+#include <Cabana_Distributor.hpp>
+
+#include <Cajita_GlobalGrid.hpp>
+#include <Cajita_GlobalMesh.hpp>
+#include <Cajita_LocalGrid.hpp>
+#include <Cajita_LocalMesh.hpp>
+
+#include <Kokkos_Core.hpp>
+
+#include <vector>
+
+namespace Cajita
+{
+
+//---------------------------------------------------------------------------//
+// Particle Grid Distributor
+//---------------------------------------------------------------------------//
+/*!
+  \brief Determine which particles need to be migrated using the grid
+
+  \tparam LocalGridType Cajita LocalGrid type.
+*/
+template <class LocalGridType>
+class ParticleGridDistributor
+{
+  public:
+    using grid_type = LocalGridType;
+    // Spatial dimension.
+    static constexpr std::size_t num_space_dim = grid_type::num_space_dim;
+
+    ParticleGridDistributor( const LocalGridType& local_grid )
+        : _local_grid( local_grid )
+    {
+    }
+
+    // Check for the number of particles that must be communicated
+    template <class PositionSliceType>
+    int migrateCount( const PositionSliceType& positions,
+                      const int minimum_halo_width )
+    {
+        using execution_space = typename PositionSliceType::execution_space;
+
+        // Check within the halo width, within the ghosted domain.
+        const auto& local_mesh =
+            Cajita::createLocalMesh<Kokkos::HostSpace>( _local_grid );
+
+        Kokkos::Array<double, num_space_dim> local_low{};
+        Kokkos::Array<double, num_space_dim> local_high{};
+        for ( std::size_t d = 0; d < num_space_dim; ++d )
+        {
+            auto dx = _local_grid.globalGrid().globalMesh().cellSize( d );
```

Comment:
This implies that a uniform mesh is being used. In hindsight, given our curvilinear grid development elsewhere I wish that Cajita only did logically uniform grids but we need to support the other grids here. For now you could simply do a static assert that the user gave you a `LocalGrid` derived form a `UniformMesh`

Reply:
Added, but I could add a non-uniform version later